src/sysdolphin/baselib/fog.c

```diff
@@ -0,0 +1,256 @@
+
+void HSD_FogInit(HSD_Fog* fog, HSD_FogDesc* desc)
+{
+    if (fog != NULL) {
+        if (desc != NULL) {
+            fog->type = desc->type;
+            fog->start = desc->start;
+            fog->end = desc->end;
+            fog->color = desc->color;
+        } else {
+            f32 sp10[6];
+            GXGetViewportv(sp10);
+            fog->type = 2;
+            fog->start = sp10[4];
+            fog->end = sp10[5];
+            fog->color.r = 0xFF;
+            fog->color.g = 0xFF;
+            fog->color.b = 0xFF;
+            fog->color.a = 0xFF;
+        }
+    }
+}
+
+HSD_FogAdj* HSD_FogAdjAlloc(void)
+{
+    HSD_FogAdj* adj = hsdNew(&hsdFogAdj);
+    adj ? (void) 0 : __assert("fog.c", 0xD6, "adj");
+    return adj;
+}
+
+HSD_FogAdj* HSD_FogAdjLoadDesc(HSD_FogAdjDesc* desc)
+{
+    HSD_FogAdj* adj = HSD_FogAdjAlloc();
+    adj ? (void) 0 : __assert("fog.c", 0xE6, "adj");
+    HSD_FogAdjInit(adj, desc);
+    return adj;
+}
+
+void HSD_FogAdjInit(HSD_FogAdj* adj, HSD_FogAdjDesc* desc)
+{
+    f32 sp10[6];
+
+    if (adj != NULL) {
+        if (desc != NULL) {
+            adj->width = desc->width;
+            adj->center = desc->center;
+            PSMTXCopy(desc->mtx, adj->mtx);
+        } else {
+            GXGetViewportv(sp10);
+            adj->width = sp10[2];
+            adj->center = adj->width / 2;
+            PSMTXIdentity(adj->mtx);
+            adj->mtx[3][0] = 0.0F;
+            adj->mtx[3][1] = 0.0F;
+            adj->mtx[3][2] = 0.0F;
+            adj->mtx[3][3] = 1.0F;
+        }
+    }
+}
+
+// HSD_FogAddAnim ?
+void func_8037DE7C(HSD_Fog* fog, HSD_AObjDesc* desc)
+{
+    if (fog != NULL) {
+        if (fog->aobj != NULL) {
+            HSD_AObjRemove(fog->aobj);
+        }
+        fog->aobj = HSD_AObjLoadDesc(desc);
+    }
+}
+
+void HSD_FogReqAnim(HSD_Fog* fog, f32 frame)
+{
+    HSD_FogReqAnimByFlags(fog, 0x7FF, frame);
+}
+
+void HSD_FogReqAnimByFlags(HSD_Fog* fog, u32 flags, f32 frame)
+{
+    if (fog == NULL || !(flags & 0x200)) {
+        return;
+    }
+    HSD_AObjReqAnim(fog->aobj, frame);
+}
+
+void HSD_FogInterpretAnim(HSD_Fog* fog)
+{
+    if (fog != NULL) {
+        HSD_AObjInterpretAnim(fog->aobj, fog, FogUpdateFunc);
+    }
+}
+
+void FogUpdateFunc(HSD_Fog* fog, u32 type, FObjData* fv)
+{
+    if (fog != NULL) {
+        switch (type) {
+        case 1:
+            fog->start = fv->fv;
+            return;
+        case 2:
+            fog->end = fv->fv;
+            return;
+        case 5:
+            fog->color.r = 255.0F * fv->fv;
+            return;
+        case 6:
+            fog->color.g = 255.0F * fv->fv;
+            return;
+        case 7:
+            fog->color.b = 255.0F * fv->fv;
+            return;
+        case 8:
+            fog->color.a = 255.0F * fv->fv;
+            return;
+        case 20:
+            if (fog->fog_adj != NULL) {
+                fog->fog_adj->center = fv->fv;
+            }
+            break;
+        }
+    }
+}
+
+inline BOOL ref_DEC(void* o)
```

Comment:
`ref_DEC` is an object.c function, let's move this there.

Reply:
Sure, there are a couple others in jobj.c that need to be moved as well. I can't move the ones containing asserts yet though, because the object.h \_\_FILE__ string is referenced from the nonmatching HSD_JObjRemoveAll's inline asm.